{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conversion complete: clinvar_converted.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "def vcf_to_csv(vcf_file, output_csv):\n",
    "    data = []\n",
    "    \n",
    "    open_func = gzip.open if vcf_file.endswith(\".gz\") else open\n",
    "    with open_func(vcf_file, 'rt') as vcf:\n",
    "        for line in vcf:\n",
    "            if line.startswith(\"#\"): \n",
    "                continue\n",
    "            cols = line.strip().split(\"\\t\")\n",
    "            chrom, pos, var_id, ref, alt = cols[:5]  \n",
    "            info = cols[7]  \n",
    "            \n",
    "            gene = \"-\"\n",
    "            significance = \"-\"\n",
    "            for item in info.split(\";\"):\n",
    "                if item.startswith(\"GENEINFO=\"):\n",
    "                    gene = item.split(\"=\")[1].split(\":\")[0]  \n",
    "                if item.startswith(\"CLNSIG=\"):\n",
    "                    significance = item.split(\"=\")[1] \n",
    "            \n",
    "            data.append([chrom, pos, var_id, ref, alt, gene, significance])\n",
    "    \n",
    "    df = pd.DataFrame(data, columns=[\"Chromosome\", \"Position\", \"ID\", \"Ref\", \"Alt\", \"Gene\", \"Clinical_Significance\"])\n",
    "    \n",
    "    df.to_csv(output_csv, index=False)\n",
    "    print(f\"Conversion complete: {output_csv}\")\n",
    "\n",
    "vcf_to_csv(r\"C:\\Users\\aashutosh kumar\\Downloads\\clinvar.vcf\\clinvar.vcf\", \"clinvar_converted.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\aashutosh kumar\\AppData\\Local\\Temp\\ipykernel_12752\\221197223.py:3: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(r\"C:\\Projects\\Gene-Mutation-Detection\\clinvar_converted.csv\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>ID</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Clinical_Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>66926</td>\n",
       "      <td>3385321</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>2205837</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>69314</td>\n",
       "      <td>3205580</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>69423</td>\n",
       "      <td>3205581</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>69581</td>\n",
       "      <td>2252161</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>69682</td>\n",
       "      <td>2396347</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>69731</td>\n",
       "      <td>3205582</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>69769</td>\n",
       "      <td>2288999</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>69995</td>\n",
       "      <td>2351346</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>924518</td>\n",
       "      <td>3388928</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>Likely_benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Position       ID Ref Alt    Gene   Clinical_Significance\n",
       "0          1     66926  3385321  AG   A   OR4F5  Uncertain_significance\n",
       "1          1     69134  2205837   A   G   OR4F5           Likely_benign\n",
       "2          1     69314  3205580   T   G   OR4F5  Uncertain_significance\n",
       "3          1     69423  3205581   G   A   OR4F5  Uncertain_significance\n",
       "4          1     69581  2252161   C   G   OR4F5  Uncertain_significance\n",
       "5          1     69682  2396347   G   A   OR4F5  Uncertain_significance\n",
       "6          1     69731  3205582   T   C   OR4F5  Uncertain_significance\n",
       "7          1     69769  2288999   T   C   OR4F5  Uncertain_significance\n",
       "8          1     69995  2351346   G   C   OR4F5  Uncertain_significance\n",
       "9          1    924518  3388928   G   C  SAMD11           Likely_benign"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "data = pd.read_csv(r\"C:\\Projects\\Gene-Mutation-Detection\\clinvar_converted.csv\")\n",
    "data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\aashutosh kumar\\AppData\\Local\\Temp\\ipykernel_12456\\1070538661.py:4: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(r\"C:\\Users\\aashutosh kumar\\Downloads\\clinvar_converted.csv\")\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "df = pd.read_csv(r\"C:\\Users\\aashutosh kumar\\Downloads\\clinvar_converted.csv\")\n",
    "df.drop(columns=[\"ID\"], inplace=True)\n",
    "\n",
    "categorical_cols = [\"Gene\", \"Ref\", \"Alt\", \"Chromosome\"]\n",
    "df[categorical_cols] = df[categorical_cols].astype(str)\n",
    "\n",
    "encoders = {}\n",
    "for col in categorical_cols:\n",
    "    encoders[col] = LabelEncoder()\n",
    "    df[col] = encoders[col].fit_transform(df[col])\n",
    "\n",
    "#1 = Pathogenic, 0 = Benign\n",
    "df[\"Clinical_Significance\"] = df[\"Clinical_Significance\"].apply(lambda x: 1 if \"Pathogenic\" in x else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Clinical_Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>66926</td>\n",
       "      <td>3845</td>\n",
       "      <td>1</td>\n",
       "      <td>10871</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>69134</td>\n",
       "      <td>0</td>\n",
       "      <td>8713</td>\n",
       "      <td>10871</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>69314</td>\n",
       "      <td>24652</td>\n",
       "      <td>8713</td>\n",
       "      <td>10871</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>69423</td>\n",
       "      <td>16297</td>\n",
       "      <td>1</td>\n",
       "      <td>10871</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome  Position    Ref   Alt   Gene  Clinical_Significance\n",
       "0           0     66926   3845     1  10871                      0\n",
       "1           0     69134      0  8713  10871                      0\n",
       "2           0     69314  24652  8713  10871                      0\n",
       "3           0     69423  16297     1  10871                      0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, classification_report\n",
    "\n",
    "\n",
    "X = df.drop(columns=[\"Clinical_Significance\"])\n",
    "y = df[\"Clinical_Significance\"]\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "clf.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = clf.predict(X_test)\n",
    "\n",
    "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
    "print(classification_report(y_test, y_pred))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
